### PR TITLE
Add stricter pool testing tools

### PR DESCRIPTION
**Pull request by CodeSandwich**
Now we have pool testing tools, which check more aspects of the contract state and allow tighter and more explicit control over what's going on.

Many of the tools are built around the design of the Buidler EVM and its mining behavior. Every call (use of a `view` or `pure` function) is done on the last block, which was mined. On the other hand every transaction (use of a non-`view` and non-`pure` funciton) first creates a new block, applies the transaction and then mines it. So we have:

```
contract.viewFn();    // executes with `block.number == 5`
contract.nonViewFn(); // executes with `block.number == 6`
contract.viewFn();    // executes with `block.number == 6`
contract.viewFn();    // executes with `block.number == 6`
contract.nonViewFn(); // executes with `block.number == 7`
contract.nonViewFn(); // executes with `block.number == 8`
```

### CODE REVIEW

**Comment by cloudhead** [comment]
I'm not sure I understand why we need two functions for each thing, eg. `Foo` and `FooOnNextBlock`, instead of something like:
```
mineBlock(1);
foo();
mineBlock(1);
bar();
baz();
...
```

**Comment by CodeSandwich** [comment]
> I'm not sure I understand why we need two functions for each thing

The `withdrawable` and `collectable` amounts keep shifting on each block. If I call `pool.withdrawable()` and get 10, next call to `pool.withdraw()` will yield just 9, because it will be executed on the next block. Or 8. Or 10. Depends on the current configuration. With `pool.withdrawableOnNextBlock()` we're always getting exactly the amount, which will be actually withdrawn.

**Review by geigerzaehler — APPROVED** [review]
:clap:

File: test/pool.test.ts [hunk 1]
```diff
@@ -31,47 +43,106 @@ async function mineBlocks(count: number): Promise<void> {
...
+    const receiverPool = senderPool.connect(receiver);
 
+    await mineBlocksUntilCycleEnd();
     // Start sending
-    await senderPool.topUp({value: 100});
+    await topUp(senderPool, 0, 100);
     await senderPool.setAmountPerBlock(1);
     await senderPool.setReceiver(receiverAddr, 1);
-    await mineBlocksUntilCycleEnd(); // 7 blocks left until cycle end
+    await mineBlocksUntilCycleEnd(); // 7 blocks left in the cycle
```
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply test/pool.test.ts]: Fixed

File: test/pool.test.ts [hunk 2]
```diff
@@ -12,6 +12,16 @@ async function addr(idx: number): Promise<string> {
   return await (await buidler.ethers.getSigners())[idx].getAddress();
 }
 
+// A helper for calling contract functions on the next block.
+// Allows calling `view` functions on the block of the next transaction, but before it.
+async function callOnNextBlock<T>(called: () => T): Promise<T> {
```
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply test/pool.test.ts]: Fixed, I've also extended the docs to explain why this function is needed in the first place.

**Comment by CodeSandwich** [comment]
@cloudhead I've added an explanation, why `callOnNextBlock` is needed in its comment.

**Pull request merged**